### PR TITLE
K8SPXC-336 : refactor pxc status

### PR DESCRIPTION
**Pull request by AlyHKafoury**
[K8SPXC-336](https://jira.percona.com/browse/K8SPXC-336)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | failed |
| storage | passed |
| recreate | passed |
| big-data | passed |
| scaling | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| upgrade | passed |
| self-healing-advanced | passed |
| limits | passed |
| upgrade-consistency | passed |
| security-context | passed |
| operator-self-healing | passed |
| demand-backup | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| scheduled-backup | failed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| storage | passed |
| big-data | passed |
| haproxy | failed |
| recreate | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| upgrade | passed |
| restore-to-encrypted-cluster | passed |
| self-healing-advanced | passed |
| limits | passed |
| upgrade-consistency | passed |
| security-context | passed |
| operator-self-healing | passed |
| demand-backup | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| scheduled-backup | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup-encrypted-with-tls | passed |

**Review by hors — CHANGES_REQUESTED** [review]
@AlyHKafoury  Did you cover all points from task https://jira.percona.com/browse/K8SPXC-336 in this  review ? e.g. what about points 1, 2 and 4. I do not see changes connected with these requirements.

File: pkg/controller/pxc/status.go [hunk 1]
```diff
@@ -206,37 +208,28 @@ func (r *ReconcilePerconaXtraDBCluster) updateStatus(cr *api.PerconaXtraDBCluste
...
 		clusterCondition = api.ClusterCondition{
 			Status:             api.ConditionTrue,
 			Type:               api.ClusterError,
 			LastTransitionTime: metav1.NewTime(time.Now()),
 		}
 		cr.Status.Status = api.AppStateError
-	case cr.Status.PXC.Status == api.AppStateInit ||
-		(cr.Spec.ProxySQL != nil && cr.Status.ProxySQL.Status == api.AppStateInit) ||
-		(cr.Spec.HAProxy != nil && cr.Status.HAProxy.Status == api.AppStateInit):
+	case clusterStatus.ClusterInit():
```
Comment (dAdAbird) [inline pkg/controller/pxc/status.go:232]: Here (in [L238](https://github.com/percona/percona-xtradb-cluster-operator/pull/563/files#diff-b246f9cb3cd35db1c2d9931e00d89aabR238)) the AppStateInit being set and then later in [L259](https://github.com/percona/percona-xtradb-cluster-operator/pull/563/files#diff-b246f9cb3cd35db1c2d9931e00d89aabR259) basically a duplication. And for that duplication, we're maintaining "inProgres" variable a doing quite a bit extra stuff. I believe setting the AppStateInit here in sewitch-case is enough

File: pkg/controller/pxc/status.go [hunk 2]
```diff
@@ -190,14 +147,59 @@ func (r *ReconcilePerconaXtraDBCluster) updateStatus(cr *api.PerconaXtraDBCluste
...
+	cr.Status.Messages = cr.Status.Messages[:0]
+	clusterStatus := clusterStatus{}
+
+	if err := r.updatePXCStatus(cr, &clusterStatus); err != nil {
+		return err
+	}
+
+	inProgres := false
+
+	if err := r.updateHAProxyStatus(cr, &inProgres, &clusterStatus); err != nil {
```
Comment (dAdAbird) [inline pkg/controller/pxc/status.go:195]: I don't quite like an idea that communication is built around changing the parameters passed to the method.
Now, these methods do a bit chaotic meaning that the do a lot of different stuff. Plus these set a lot of states. It's too unclear.

**Review by dAdAbird — CHANGES_REQUESTED** [review]
See comments. 
Plus:
1. As @hors mentioned, there is nothing on p. 1 , 2 & 4.
2. There a lot of moving parts and we're constantly bumping into bugs here so I'd like to see unit tests for statuses.

**Comment by delgod** [comment]
will be reopened when finished

**Pull request closed without merging**